src/wasm/asm-wasm-builder.cc

```diff
@@ -319,21 +368,61 @@ class AsmWasmBuilderImpl : public AstVisitor {
 
   void VisitBinaryOperation(BinaryOperation* expr) {
     switch (expr->op()) {
+      // change to use xmacros and deal with signed/unsigned
       case Token::BIT_OR:
-        current_function_builder_->AppendCode(kExprI32Ior, false);
+        if (TypeOf(expr) == kAstI32) {
+          current_function_builder_->AppendCode(kExprI32Ior, false);
+        } else {
+          UNREACHABLE();
+        }
         break;
       case Token::ADD:
-        current_function_builder_->AppendCode(kExprI32Add, false);
+        if (TypeOf(expr) == kAstI32) {
+          current_function_builder_->AppendCode(kExprI32Add, false);
+        } else if (TypeOf(expr) == kAstF32){
+          current_function_builder_->AppendCode(kExprF32Add, false);
+        } else if (TypeOf(expr) == kAstF64){
+          current_function_builder_->AppendCode(kExprF64Add, false);
+        } else {
+          UNREACHABLE();
+        }
         break;
-      default:
-        DCHECK(false);
+      case Token::MUL:
+        if (TypeOf(expr) == kAstI32) {
+          current_function_builder_->AppendCode(kExprI32Mul, false);
+        } else if (TypeOf(expr) == kAstF32){
+          current_function_builder_->AppendCode(kExprF32Mul, false);
+        } else if (TypeOf(expr) == kAstF64){
+          current_function_builder_->AppendCode(kExprF64Mul, false);
+        } else {
+          UNREACHABLE();
+        }
         break;
+      default:
+        UNREACHABLE();
     }
     RECURSE(Visit(expr->left()));
     RECURSE(Visit(expr->right()));
   }
 
   void VisitCompareOperation(CompareOperation* expr) {
+    switch(expr->op()) {
+      // change to use xmacros and deal with signed/unsigned
```

Comment:
There's gonna be a lot of code like this with a cascade of if's over the WASM local type. It'll be cleaner to pull these out to utility functions that can be tested independently.